pkg/odo/cli/component/common_push.go

```diff
@@ -330,3 +332,71 @@ func filterIgnores(filesChanged, filesDeleted, absIgnoreRules []string) (filesCh
 	}
 	return filesChangedFiltered, filesDeletedFiltered
 }
+
+// retrieveDefaultComponentNamespace tries to retrieve the current active namespace
+// to set as a default namespace
+func retrieveKubernetesDefaultNamespace() (string, error) {
+	// Get current active namespace
+	client, err := kclient.New()
+	if err != nil {
+		return "", err
+	}
+	return client.Namespace, nil
+}
+
+// checkNamespaceDefaultIfNotAvailable checks to see if a namespace has been passed in, if not, we will default
+// to the standard kubernetes namespace.
+func checkCmdNamespaceDefaultIfNotAvailable(cmd *cobra.Command) (string, error) {
+	var componentNamespace string
+
+	// For "odo create" check to see if --project has been passed.
+	if cmd.Flags().Changed("project") {
+		componentNamespace, err := cmd.Flags().GetString("project")
+		if err != nil {
+			return componentNamespace, err
+		}
+	} else if cmd.Flags().Changed("namespace") {
+		// For "odo push" check to see if project has been passed
+		componentNamespace, err := cmd.Flags().GetString("namespace")
+		if err != nil {
+			return componentNamespace, err
+		}
+	}
+
+	componentNamespace, err := retrieveKubernetesDefaultNamespace()
+	if err != nil {
+		return "", err
+	}
+
+	return componentNamespace, nil
+}
+
+// gatherName parses the Devfile retrieves an appropriate name in two ways.
+// 1. If metadata.name exists, we use it
+// 2. If metadata.name does NOT exist, we use the folder name where the devfile.yaml is located
+func gatherName(devfilePath string) (string, error) {
+
+	// Parse the Devfile
+	devObj, err := devfile.Parse(devfilePath)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to parse devfile")
+	}
+	metadata := devObj.Data.GetMetadata()
+
+	// 1. Use metadata.name if it exists
+	if metadata.Name != "" {
+
+		// Make sure that we remove any "dashes" in the name as they will not work
+		// correctly with Kubernetes component deployment
+		return strings.Replace(metadata.Name, "-", "", -1), nil
+	}
+
+	// 2. Use the folder name as a last resort if nothing else exists
+	sourcePath, err := util.GetAbsPath(devfilePath)
```

Comment:
I tried devfile 1.0 as well as 2.0. In both the cases, `odo push` fails if there's not `metadata.name` value. Does this condition really get called at some point? :thinking: